lib/manager/terraform/artifacts.ts

```diff
@@ -0,0 +1,78 @@
+import { logger } from '../../logger';
+import { ExecOptions, exec } from '../../util/exec';
+import {
+  deleteLocalFile,
+  getSiblingFileName,
+  readLocalFile,
+  writeLocalFile,
+} from '../../util/fs';
+import { UpdateArtifact, UpdateArtifactsResult } from '../common';
+
+const lockFile = '.terraform.lock.hcl';
+
+async function terraformLockUpdate(): Promise<void> {
+  await deleteLocalFile(lockFile);
+  const cmd = 'terraform providers lock';
+
+  const execOptions: ExecOptions = {
+    docker: {
+      image: 'renovate/terraform',
+    },
+  };
+  await exec(cmd, execOptions);
+}
+
+export async function updateArtifacts({
+  packageFileName,
+  updatedDeps,
+  newPackageFileContent,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`terraform.updateArtifacts(${packageFileName})`);
+
+  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+
+  if (
+    !isLockFileMaintenance &&
+    (updatedDeps === undefined || updatedDeps.length < 1)
+  ) {
+    logger.debug('No updated terraform deps - returning null');
+    return null;
+  }
+
+  const lockFileName = getSiblingFileName(packageFileName, lockFile);
+  const existingLockFileContent = await readLocalFile(lockFileName);
+  if (!existingLockFileContent) {
+    logger.debug('No .terraform.lock.hcl found');
+    return null;
+  }
```

Comment:
So this get potencially run multiple times or never too.

1. never run when there is no update in lockfile folder
2. multiple runs when there are multiple updates in that folder

@rarkins how far are we with the seperate lock files properties? Maybe we can add the lockfile as seperate packageFile with only lockfile update? 🤔

Reply:
I think in theory that we need to run a lock file update (a) if any deps change, or (b) if we upgrade the terraform version. Not sure though whether treating the lockfile as its own package file makes sense or fits into the above requirements

Reply:
If we incorporate following request https://github.com/renovatebot/renovate/issues/7895#issuecomment-754942081 we need to run it on the default branch too and if it has changed create a new branch and  PR. So basically: 
1. When a provider is updated
2. When Terraform is updated 
3. On the master/default branch to update the version in the lockFile

Reply:
@secustor this is what we call lockfile maintenance update. So in update artifacts you get a lockfile maintenance update type. Which should only update the lockfile to latest versions in range.